chain/update.go

```diff
@@ -0,0 +1,150 @@
+// it doesn't exist we initialize one with the given state.
+func (cu *Update) ContractUpdate(fcid types.FileContractID, state api.ContractState) *ContractUpdate {
+	_, ok := cu.ContractUpdates[fcid]
+	if !ok {
+		cu.ContractUpdates[fcid] = &ContractUpdate{State: state}
+	}
+	return cu.ContractUpdates[fcid]
+}
+
+// AddEvents is called with all relevant events added in the update.
+func (cu *Update) AddEvents(events []wallet.Event) error {
+	for _, event := range events {
+		cu.WalletEventUpdates = append(cu.WalletEventUpdates, WalletEventUpdate{
+			Addition: true,
+			Event:    event,
+		})
+	}
+	return nil
+}
+
+// AddSiacoinElements is called with all new siacoin elements in the
+// update. Ephemeral siacoin elements are not included.
+func (cu *Update) AddSiacoinElements(ses []wallet.SiacoinElement) error {
+	for _, se := range ses {
+		cu.WalletOutputUpdates[se.ID] = WalletOutputUpdate{
+			Addition: true,
+			ID:       se.ID,
+			Element:  se,
+		}
+	}
+	return nil
+}
+
+// RemoveSiacoinElements is called with all siacoin elements that were
+// spent in the update.
+func (cu *Update) RemoveSiacoinElements(ids []types.SiacoinOutputID) error {
+	for _, id := range ids {
+		cu.WalletOutputUpdates[types.Hash256(id)] = WalletOutputUpdate{
+			Addition: false,
+			ID:       types.Hash256(id),
+		}
+	}
+	return nil
+}
+
+// WalletStateElements returns all state elements in the database. It is used
+// to update the proofs of all state elements affected by the update.
+func (cu *Update) WalletStateElements() (elements []types.StateElement, _ error) {
+	for id, el := range cu.WalletOutputUpdates {
+		elements = append(elements, types.StateElement{
+			ID:          id,
+			LeafIndex:   el.Element.LeafIndex,
+			MerkleProof: el.Element.MerkleProof,
+		})
+	}
+	return
+}
+
+// UpdateStateElements updates the proofs of all state elements affected by the
+// update.
+func (cu *Update) UpdateStateElements(elements []types.StateElement) error {
+	for _, se := range elements {
+		curr := cu.WalletOutputUpdates[se.ID]
+		curr.Element.MerkleProof = se.MerkleProof
+		curr.Element.LeafIndex = se.LeafIndex
+		cu.WalletOutputUpdates[se.ID] = curr
+	}
+	return nil
+}
+
+// RevertIndex is called with the chain index that is being reverted. Any events
+// and siacoin elements that were created by the index should be removed.
+func (cu *Update) RevertIndex(index types.ChainIndex) error {
+	// remove any events that were added in the reverted block
+	filtered := cu.WalletEventUpdates[:0]
+	for i := range cu.WalletEventUpdates {
+		if cu.WalletEventUpdates[i].Event.Index != index {
+			filtered = append(filtered, cu.WalletEventUpdates[i])
+		}
+	}
+	cu.WalletEventUpdates = filtered
+
+	// remove any siacoin elements that were added in the reverted block
+	for id, el := range cu.WalletOutputUpdates {
+		if el.Element.Index == index {
+			delete(cu.WalletOutputUpdates, id)
+		}
+	}
+
+	return nil
```

Comment:
This `RevertIndex` won't revert an index that is not currently in `Update`.
e.g. when `sync` is triggered by `OnReorg`, we create a new update with `NewChainUpdate` and then the first update is a revert. Which is usually going to be the case.

So we probably need a field like `RevertIndexUpdates` to remember these and then apply them to the database.

Reply:
I updated this, I know we can batch delete when applying these reverts but I did not bother. I think looping them is fine?

Reply:
Reverting them one-by-one is fine yeah. Shouldn't happen too often anyway.